src/script_opt/ZAM/Expr.cc

```diff
@@ -78,6 +82,56 @@ const ZAMStmt ZAMCompiler::CompileAppendToExpr(const AppendToExpr* e) {
     return n2 ? AppendToVV(n1, n2) : AppendToVC(n1, cc);
 }
 
+const ZAMStmt ZAMCompiler::CompileAdd(const AggrAddExpr* e) {
+    auto op = e->GetOp1();
+    auto aggr = op->GetOp1()->AsNameExpr();
+    auto index_list = op->GetOp2();
+
+    if ( index_list->Tag() != EXPR_LIST )
+        reporter->InternalError("non-list in \"add\"");
+
+    auto indices = index_list->AsListExprPtr();
+    auto& exprs = indices->Exprs();
+
+    if ( exprs.length() == 1 ) {
+        auto e1 = exprs[0];
+        if ( e1->Tag() == EXPR_NAME )
+            return AddStmt1VV(aggr, e1->AsNameExpr());
+        else
+            return AddStmt1VC(aggr, e1->AsConstExpr());
+    }
+
+    return AddStmtVO(aggr, BuildVals(indices));
+}
+
+const ZAMStmt ZAMCompiler::CompileDel(const AggrDelExpr* e) {
+    auto op = e->GetOp1();
+
+    if ( op->Tag() == EXPR_NAME ) {
+        auto n = op->AsNameExpr();
+
+        if ( n->GetType()->Tag() == TYPE_TABLE )
+            return ClearTableV(n);
+        else
+            return ClearVectorV(n);
+    }
+
+    auto aggr = op->GetOp1()->AsNameExpr();
+
+    if ( op->Tag() == EXPR_FIELD ) {
+        int field = op->AsFieldExpr()->Field();
+        return DelFieldVi(aggr, field);
+    }
+
+    auto index_list = op->GetOp2();
+
+    if ( index_list->Tag() != EXPR_LIST )
+        reporter->InternalError("non-list in \"delete\"");
+
+    auto internal_ind = std::unique_ptr<OpaqueVals>(BuildVals(index_list->AsListExprPtr()));
```

Comment:
Is this `unique_ptr` just because `DelTableV0` doesn't destroy the pointer?

Reply:
Yes. `DelTableV0` is auto-generated by `gen-zam`, and those methods work on raw pointers.

(Note, this code is old, it just shows up here because it was moved from `src/script_opt/ZAM/Stmt.cc` to here.)